docs/source/reference-io.rst

```diff
@@ -1,26 +1,171 @@
+.. currentmodule:: trio
+
 I/O in Trio
 ===========
 
+.. _abstract-stream-api:
+
+The abstract Stream API
+-----------------------
+
+Trio provides a set of abstract base classes that define a standard
+interface for unidirectional and bidirectional byte streams.
+
+Why is this useful? Because it lets you write generic protocol
+implementations that can work over arbitrary transports, and easily
+create complex transport configurations. Here's some examples:
+
+* :class:`trio.SocketStream` wraps a raw socket (like a TCP connection
+  over the network), and converts it to the standard stream interface.
+
+* :class:`trio.ssl.SSLStream` is a "stream adapter" that can take any
+  object that implements the :class:`trio.abc.Stream` interface, and
+  convert it into an encrypted stream. In trio the standard way to
+  speak SSL over the network is to wrap an
+  :class:`~trio.ssl.SSLStream` around a :class:`~trio.SocketStream`.
+
+* If you spawn a subprocess then you can get a
+  :class:`~trio.abc.SendStream` that lets you write to its stdin, and
+  a :class:`~trio.abc.ReceiveStream` that lets you read from its
+  stdout. If for some reason you wanted to speak SSL to a subprocess,
+  you could use a :class:`StapledStream` to combine its stdin/stdout
+  into a single bidirectional :class:`~trio.abc.Stream`, and then wrap
+  that in an :class:`~trio.ssl.SSLStream`::
+
+     ssl_context = trio.ssl.create_default_context()
+     ssl_context.check_hostname = False
+     s = SSLStream(StapledStream(process.stdin, process.stdout), ssl_context)
+
+  [Note: subprocess support is not implemented yet, but that's the
```

Comment:
I think I'll leave this as is for now, it's not really a call-out-box kind of note, more like a whispered parenthetical.

If that makes any sense at all, which it probably doesn't.